transform/src/lib.rs

```diff
@@ -78,35 +67,234 @@ pub async fn initialize_staged(
     // - No runtime table operations
     // - No reference type globals
     // - Each module instantiated at most once
+    //
+    // instrumentation keeps track of all of the state which will be gathered from the instrumented
+    // component.
+    let (instrumented_component, instrumentation) = instrument(component_stage1)?;
 
-    let copy_component_section = |section, component: &[u8], result: &mut Component| {
-        if let Some((id, range)) = section {
-            result.section(&RawSection {
-                id,
-                data: &component[range],
-            });
+    Validator::new().validate_all(&instrumented_component)?;
+
+    // A component runtime will instantiate the component and run its component init function.
+    let mut invoker = initialize(instrumented_component).await?;
+
+    // The Invoker interface is used to extract the values instrumentation provided into a
+    // measurement.
+    let measurement = instrumentation.measure(&mut invoker).await?;
+
+    // Finally, create a new component by applying the measurement (contents of all globals and memory) to the component.
+    // The resulting component will identical to the original except with all mutable globals initialized to
+    // the snapshoted values, with all data sections and start functions removed, and with a single active data
+    // section added containing the memory snapshot.
+    apply(
+        measurement,
+        component_stage1,
+        component_stage2_and_map_module_index,
+    )
+}
+
+struct MemoryInfo {
+    module_index: u32,
+    export_name: String,
+    ty: wasmparser::MemoryType,
+}
+type GlobalMap<T> = HashMap<u32, HashMap<u32, T>>;
+#[derive(Debug)]
+enum GlobalExport {
```

Comment:
Nit, I'd recommend using named fields for the payloads of these variants so the reader doesn't need to search elsewhere to discover what they represent.